### ISSUE TITLE
Hdfs NameNode Service not headless 

### ISSUE THREAD

**Issue by epelund** [Gradiant/charts#3]
Hi,

Is there any special reason for the missing "clusterIP: None" in [Hdfs NameNode Service](https://github.com/Gradiant/charts/blob/8aeb218d640292e9495d9bc7e57303e404c1027b/charts/hdfs/templates/hdfs-nn-svc.yaml#L10) compared to the one in [helm stable](https://github.com/helm/charts/blob/master/stable/hadoop/templates/hdfs-nn-svc.yaml#L18)? If the service type is ClusterIP the datanode IP will be masqued by the kubernetes iptables and the namenode will [register](https://github.com/apache/hadoop/blob/branch-2.7.7/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/DatanodeManager.java#L842l) it by the worker node's IP address. This will fail in for example a single node MicroK8s with multiple datanodes as the namenode will overwrite the registration and only one datanode will be registered at a time. I had some headache over this;-. 

BR
/Peter

**Comment by cgiraldo** [comment]
Hi,

It is true that hdfs namenode service is not headless, because we want the hdfs-namenode to be registered in kubernetes DNS and access with a single service IP. It is true we must remove the initial comment of the file saying it is a headless service.

On the other hand, the [Hdfs Datanode Service](https://github.com/Gradiant/charts/blob/8aeb218d640292e9495d9bc7e57303e404c1027b/charts/hdfs/templates/hdfs-dn-svc.yaml#L14) is headless to avoid the problem you mention.

We have tested this chart with the defaults (datanode instances=3) with a single-node k3s cluster with success.
Can you test it in your environment and provide more details if your datanodes do not register?

Kind regards

**Comment by epelund** [comment]
They register but the registrations are overwritten as the NameNode register the DataNodes by the source IP which in my understanding is masqued to Node IP by the ip filters defined by the kube-proxy for the NameNode ClusterIP service. When using headless those filters are gone and the DataNodes correctly registers by their pod IP which is visible in the NameNode UI. Otherwise I can only see one datanode at a time in the UI with the same IP (Node) but different hostnames. Are you running kube-proxy in iptables mode?

Great job with the pnda charts btw!

**Comment by epelund** [comment]
What happens with ClusterIP on the NameNode service:

<pre>
$ kubectl exec -it hdfs-test-namenode-0 -c namenode -- bash -c "hdfs dfsadmin -report | grep -A 3 Live"
Live datanodes (1):

Name: 10.1.96.1:50010 (10.1.96.1)
Hostname: hdfs-test-datanode-2.hdfs-test-datanode.pducc-nwst-dev.svc.cluster.local
</pre>

The the IP of the datanode is the Node IP and the hostname of the one registered datanode keeps changing between hdfs-test-datanode-0|1|2.

If I try to create a file it fails:

<pre>
$ kubectl exec -it hdfs-test-namenode-0 -c namenode -- bash -c "hadoop fs -mkdir /test; hadoop fs -put ./entrypoint.sh /test/test.txt; hadoop fs -ls /test/"
20/03/20 10:42:47 WARN hdfs.DFSClient: DataStreamer Exception
org.apache.hadoop.ipc.RemoteException(java.io.IOException): File /test/test.txt._COPYING_ could only be replicated to 1 nodes instead of minReplication (=3).  There are 1 datanode(s) running and no node(s) are excluded in this operation.
	at org.apache.hadoop.hdfs.server.blockmanagement.BlockManager.chooseTarget4NewBlock(BlockManager.java:1625)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getNewBlockTargets(FSNamesystem.java:3135)
	at org.apache.hadoop.hdfs.server.namenode.FSNamesystem.getAdditionalBlock(FSNamesystem.java:3059)
	at org.apache.hadoop.hdfs.server.namenode.NameNodeRpcServer.addBlock(NameNodeRpcServer.java:725)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolServerSideTranslatorPB.addBlock(ClientNamenodeProtocolServerSideTranslatorPB.java:493)
	at org.apache.hadoop.hdfs.protocol.proto.ClientNamenodeProtocolProtos$ClientNamenodeProtocol$2.callBlockingMethod(ClientNamenodeProtocolProtos.java)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Server$ProtoBufRpcInvoker.call(ProtobufRpcEngine.java:616)
	at org.apache.hadoop.ipc.RPC$Server.call(RPC.java:982)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2217)
	at org.apache.hadoop.ipc.Server$Handler$1.run(Server.java:2213)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1762)
	at org.apache.hadoop.ipc.Server$Handler.run(Server.java:2213)

	at org.apache.hadoop.ipc.Client.call(Client.java:1476)
	at org.apache.hadoop.ipc.Client.call(Client.java:1413)
	at org.apache.hadoop.ipc.ProtobufRpcEngine$Invoker.invoke(ProtobufRpcEngine.java:229)
	at com.sun.proxy.$Proxy10.addBlock(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.addBlock(ClientNamenodeProtocolTranslatorPB.java:418)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:191)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:102)
	at com.sun.proxy.$Proxy11.addBlock(Unknown Source)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.locateFollowingBlock(DFSOutputStream.java:1603)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.nextBlockOutputStream(DFSOutputStream.java:1388)
	at org.apache.hadoop.hdfs.DFSOutputStream$DataStreamer.run(DFSOutputStream.java:554)
put: File /test/test.txt._COPYING_ could only be replicated to 1 nodes instead of minReplication (=3).  There are 1 datanode(s) running and no node(s) are excluded in this operation.
</pre>

The suspected iptable rules:

<pre>
$ sudo iptables -L -v -t nat | grep hdfs
    0     0 KUBE-MARK-MASQ  tcp  --  any    any    !10.152.183.0/24      10.152.183.120       /* pducc-nwst-dev/hdfs-test-httpfs:httpfs cluster IP */ tcp dpt:14000
    0     0 KUBE-SVC-RE4ULTPW47HKGMBS  tcp  --  any    any     anywhere             10.152.183.120       /* pducc-nwst-dev/hdfs-test-httpfs:httpfs cluster IP */ tcp dpt:14000
   31  1860 KUBE-MARK-MASQ  tcp  --  any    any    !10.152.183.0/24      10.152.183.219       /* pducc-nwst-dev/hdfs-test-namenode:dfs cluster IP */ tcp dpt:8020
   31  1860 KUBE-SVC-LBLATQ4QUI5SOPVR  tcp  --  any    any     anywhere             10.152.183.219       /* pducc-nwst-dev/hdfs-test-namenode:dfs cluster IP */ tcp dpt:8020
    0     0 KUBE-MARK-MASQ  tcp  --  any    any    !10.152.183.0/24      10.152.183.36        /* pducc-nwst-dev/hdfs-test-namenode-exporter:metrics cluster IP */ tcp dpt:freeciv
    0     0 KUBE-SVC-3366URZYQY2LUP63  tcp  --  any    any     anywhere             10.152.183.36        /* pducc-nwst-dev/hdfs-test-namenode-exporter:metrics cluster IP */ tcp dpt:freeciv
    3   180 KUBE-MARK-MASQ  tcp  --  any    any    !10.152.183.0/24      10.152.183.219       /* pducc-nwst-dev/hdfs-test-namenode:webhdfs cluster IP */ tcp dpt:50070
    3   180 KUBE-SVC-Y7IMQRD3WZ4NFJ33  tcp  --  any    any     anywhere             10.152.183.219       /* pducc-nwst-dev/hdfs-test-namenode:webhdfs cluster IP */ tcp dpt:50070
</pre>

Services and Pods:

<pre>
$ kubectl get svc -o wide
NAME                          TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)              AGE   SELECTOR
hdfs-test-datanode            ClusterIP   None             <none>        50075/TCP            29m   app.kubernetes.io/component=datanode,app.kubernetes.io/instance=hdfs-test,app.kubernetes.io/name=hdfs
hdfs-test-httpfs              ClusterIP   10.152.183.120   <none>        14000/TCP            29m   app.kubernetes.io/component=httpfs,app.kubernetes.io/instance=hdfs-test,app.kubernetes.io/name=hdfs
hdfs-test-namenode            ClusterIP   10.152.183.219   <none>        8020/TCP,50070/TCP   29m   app.kubernetes.io/component=namenode,app.kubernetes.io/instance=hdfs-test,app.kubernetes.io/name=hdfs
hdfs-test-namenode-exporter   ClusterIP   10.152.183.36    <none>        5556/TCP             29m   app.kubernetes.io/component=namenode,app.kubernetes.io/instance=hdfs-test,app.kubernetes.io/name=hdfs
</pre>

<pre>
$ kubectl get po -o wide
NAME                                        READY   STATUS    RESTARTS   AGE   IP            NODE        NOMINATED NODE   READINESS GATES
hdfs-test-datanode-0                        1/1     Running   0          35m   10.1.96.179   ts-ro-081   <none>           <none>
hdfs-test-datanode-1                        1/1     Running   0          35m   10.1.96.180   ts-ro-081   <none>           <none>
hdfs-test-datanode-2                        1/1     Running   0          35m   10.1.96.181   ts-ro-081   <none>           <none>
hdfs-test-httpfs-5b76466786-5fdkw           1/1     Running   0          35m   10.1.96.177   ts-ro-081   <none>           <none>
hdfs-test-namenode-0                        2/2     Running   1          35m   10.1.96.178   ts-ro-081   <none>           <none>
</pre>

**Comment by cgiraldo** [comment]
Thanks for the detailled info.
I have just installed microk8s and I can replicate the problem. I will investigate and try to fix it.

With k3s this doesn't happen.
Here are  the outputs in my k3s environment:
```
kubectl -n hdfs exec -it hdfs-namenode-0 -c namenode -- bash -c "hdfs dfsadmin -report | grep -A 3 Live"
Live datanodes (3):

Name: 10.42.0.21:50010 (hdfs-datanode-0.hdfs-datanode.hdfs.svc.cluster.local)
Hostname: hdfs-datanode-0.hdfs-datanode.hdfs.svc.cluster.local
```
Success creating file:
```
 kubectl -n hdfs exec -it hdfs-namenode-0 -c namenode -- bash -c "hadoop fs -mkdir /test; hadoop fs -put ./entrypoint.sh /test/test.txt; hadoop fs -ls /test/"
Found 1 items
-rw-r--r--   3 root supergroup       2392 2020-03-20 11:41 /test/test.txt
```
Ip tables:
```
 0     0 KUBE-MARK-MASQ  tcp  --  any    any    !pepino/16            10.43.144.173        /* hdfs/hdfs-httpfs:httpfs cluster IP */ tcp dpt:14000
    0     0 KUBE-SVC-5MDNIDQFDCIO3VB7  tcp  --  any    any     anywhere             10.43.144.173        /* hdfs/hdfs-httpfs:httpfs cluster IP */ tcp dpt:14000
    0     0 KUBE-MARK-MASQ  tcp  --  any    any    !pepino/16            10.43.155.0          /* hdfs/hdfs-namenode:webhdfs cluster IP */ tcp dpt:50070
    3   180 KUBE-SVC-7TRCTMGMNYXPBLKW  tcp  --  any    any     anywhere             10.43.155.0          /* hdfs/hdfs-namenode:webhdfs cluster IP */ tcp dpt:50070
    0     0 KUBE-MARK-MASQ  tcp  --  any    any    !pepino/16            10.43.50.179         /* hdfs/hdfs-namenode-exporter:metrics cluster IP */ tcp dpt:freeciv
    0     0 KUBE-SVC-F7OZGHZ5673MIAWJ  tcp  --  any    any     anywhere             10.43.50.179         /* hdfs/hdfs-namenode-exporter:metrics cluster IP */ tcp dpt:freeciv
    0     0 KUBE-MARK-MASQ  tcp  --  any    any    !pepino/16            10.43.155.0          /* hdfs/hdfs-namenode:dfs cluster IP */ tcp dpt:8020
    6   360 KUBE-SVC-HCY55RWDJZHE7BS6  tcp  --  any    any     anywhere             10.43.155.0          /* hdfs/hdfs-namenode:dfs cluster IP */ tcp dpt:8020

```
Services and Pods:
```
kubectl -n hdfs get svc -o wide
NAME                     TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)              AGE   SELECTOR
hdfs-datanode            ClusterIP   None            <none>        50075/TCP            11m   app.kubernetes.io/component=datanode,app.kubernetes.io/instance=hdfs,app.kubernetes.io/name=hdfs
hdfs-namenode            ClusterIP   10.43.155.0     <none>        8020/TCP,50070/TCP   11m   app.kubernetes.io/component=namenode,app.kubernetes.io/instance=hdfs,app.kubernetes.io/name=hdfs
hdfs-httpfs              ClusterIP   10.43.144.173   <none>        14000/TCP            11m   app.kubernetes.io/component=httpfs,app.kubernetes.io/instance=hdfs,app.kubernetes.io/name=hdfs
hdfs-namenode-exporter   ClusterIP   10.43.50.179    <none>        5556/TCP             11m   app.kubernetes.io/component=namenode,app.kubernetes.io/instance=hdfs,app.kubernetes.io/name=hdfs

```
```
kubectl -n hdfs get pods -o wide
NAME                          READY   STATUS    RESTARTS   AGE   IP           NODE     NOMINATED NODE   READINESS GATES
hdfs-httpfs-d84f7c8bd-b48wp   1/1     Running   0          11m   10.42.0.20   pepino   <none>           <none>
hdfs-namenode-0               2/2     Running   2          11m   10.42.0.22   pepino   <none>           <none>
hdfs-datanode-0               1/1     Running   0          11m   10.42.0.21   pepino   <none>           <none>
hdfs-datanode-1               1/1     Running   0          10m   10.42.0.23   pepino   <none>           <none>
hdfs-datanode-2               1/1     Running   0          10m   10.42.0.24   pepino   <none>           <none>

```

**Comment by cgiraldo** [comment]
Thanks! 

I change namenode service to headless and now It is working with microk8s.
I also added a hdfs clusterIP service for clients to access hdfs-namenode thoruh single clusterIP.

You can find the fixed hdfs chart in release [hdfs-0.1.4](https://github.com/Gradiant/charts/releases/tag/hdfs-0.1.4)

**Comment by epelund** [comment]
The chains involved looks like this with mark/match for the masqureade behavior on MicroK8s. Maybe a difference on the machine running the k3s environment? 
<pre>
$ sudo iptables -L -v -t nat | egrep -A 3 'Chain (KUBE-MARK-MASQ|KUBE-POSTROUTING) .*'
Chain KUBE-MARK-MASQ (36 references)
 pkts bytes target     prot opt in     out     source               destination         
  876 88718 MARK       all  --  any    any     anywhere             anywhere             MARK or 0x4000

--
Chain KUBE-POSTROUTING (1 references)
 pkts bytes target     prot opt in     out     source               destination         
  873 88538 MASQUERADE  all  --  any    any     anywhere             anywhere             /* kubernetes service traffic requiring SNAT */ mark match 0x4000/0x4000
</pre>
